packages/@lwc/engine-core/src/framework/base-lightning-element.ts

```diff
@@ -509,8 +534,17 @@ LightningElement.prototype = {
         return getClassList(elm);
     },
 
-    get template(): ShadowRoot {
+    get template(): ShadowRoot | null {
         const vm = getAssociatedVM(this);
+
+        if (!hasShadow(vm)) {
+            if (process.env.NODE_ENV !== 'production') {
```

Comment:
Swap the 2 block conditions. By having the `process.env.NODE_ENV !== 'production'` running first, the `hasShadow` will never be invoked even in production.

```suggestion
        if (process.env.NODE_ENV !== 'production') {
            if (!hasShadow(vm)) {
```